File: backend/src/main/java/com/contentplatform/backend/interfaces/web/controller/AdminMenuController.java
package com.contentplatform.backend.interfaces.web.controller;

import com.contentplatform.backend.application.service.PageMenuService;
import com.contentplatform.backend.domain.value.MenuStatus;
import com.contentplatform.backend.domain.value.ServicePermission;
import com.contentplatform.backend.interfaces.web.SecurityUtils;
import com.contentplatform.backend.interfaces.web.request.MenuItemUpsertRequest;
import com.contentplatform.backend.interfaces.web.request.MenuItemsLayoutRequest;
import com.contentplatform.backend.interfaces.web.request.MenuStatusRequest;
import com.contentplatform.backend.interfaces.web.request.MenuUpsertRequest;
import com.contentplatform.backend.interfaces.web.response.MenuContentCandidateResponse;
import com.contentplatform.backend.interfaces.web.response.MenuResponse;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/v1/admin/menus")
public class AdminMenuController {
    private final PageMenuService pageMenuService;

    public AdminMenuController(PageMenuService pageMenuService) {
        this.pageMenuService = pageMenuService;
    }

    @PostMapping
    public ResponseEntity<MenuResponse> create(@Valid @RequestBody MenuUpsertRequest request) {
        SecurityUtils.requireServicePermission(ServicePermission.MENUS_MANAGE);
        SecurityUtils.requireApplicationAccess(request.getApplicationId());
        return ResponseEntity.ok(pageMenuService.createMenu(request));
    }

    @PutMapping("/{id}")
    public ResponseEntity<MenuResponse> update(@PathVariable String id, @Valid @RequestBody MenuUpsertRequest request) {
        String applicationId = pageMenuService.getMenuApplicationId(id);
        SecurityUtils.requireServicePermission(ServicePermission.MENUS_MANAGE);
        SecurityUtils.requireApplicationAccess(applicationId);
        return ResponseEntity.ok(pageMenuService.updateMenu(id, request));
    }

    @PatchMapping("/{id}/status")
    public ResponseEntity<MenuResponse> changeStatus(@PathVariable String id, @Valid @RequestBody MenuStatusRequest request) {
        String applicationId = pageMenuService.getMenuApplicationId(id);
        SecurityUtils.requireServicePermission(ServicePermission.MENUS_MANAGE);
        SecurityUtils.requireApplicationAccess(applicationId);
        return ResponseEntity.ok(pageMenuService.changeMenuStatus(id, request.getStatus()));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Map<String, Boolean>> delete(@PathVariable String id) {
        String applicationId = pageMenuService.getMenuApplicationId(id);
        SecurityUtils.requireServicePermission(ServicePermission.MENUS_MANAGE);
        SecurityUtils.requireApplicationAccess(applicationId);
        pageMenuService.deleteMenu(id);
        return ResponseEntity.ok(Map.of("ok", true));
    }

    @GetMapping("/{id}")
    public ResponseEntity<MenuResponse> get(@PathVariable String id) {
        String applicationId = pageMenuService.getMenuApplicationId(id);
        SecurityUtils.requireServicePermission(ServicePermission.MENUS_MANAGE);
        SecurityUtils.requireApplicationAccess(applicationId);
        return ResponseEntity.ok(pageMenuService.getMenu(id));
    }

    @GetMapping("/{id}/published-content")
    public ResponseEntity<List<MenuContentCandidateResponse>> publishedContent(@PathVariable String id) {
        String applicationId = pageMenuService.getMenuApplicationId(id);
        SecurityUtils.requireServicePermission(ServicePermission.MENUS_MANAGE);
        SecurityUtils.requireApplicationAccess(applicationId);
        return ResponseEntity.ok(pageMenuService.listPublishedContentCandidates(id));
    }

    @PostMapping("/{id}/sync-published")
    public ResponseEntity<MenuResponse> syncPublished(@PathVariable String id) {
        String applicationId = pageMenuService.getMenuApplicationId(id);
        SecurityUtils.requireServicePermission(ServicePermission.MENUS_MANAGE);
        SecurityUtils.requireApplicationAccess(applicationId);
        return ResponseEntity.ok(pageMenuService.syncPublishedContent(id));
    }

    @GetMapping
    public ResponseEntity<List<MenuResponse>> list(@RequestParam String applicationId,
                                                   @RequestParam(required = false) String languageCode,
                                                   @RequestParam(required = false) MenuStatus status) {
        SecurityUtils.requireServicePermission(ServicePermission.MENUS_MANAGE);
        SecurityUtils.requireApplicationAccess(applicationId);
        return ResponseEntity.ok(pageMenuService.listMenus(applicationId, languageCode, status));
    }

    @PostMapping("/{id}/items")
    public ResponseEntity<MenuResponse> addItem(@PathVariable String id, @Valid @RequestBody MenuItemUpsertRequest request) {
        String applicationId = pageMenuService.getMenuApplicationId(id);
        SecurityUtils.requireServicePermission(ServicePermission.MENUS_MANAGE);
        SecurityUtils.requireApplicationAccess(applicationId);
        return ResponseEntity.ok(pageMenuService.addMenuItem(id, request));
    }

    @PutMapping("/{id}/items/layout")
    public ResponseEntity<MenuResponse> updateItemsLayout(@PathVariable String id, @RequestBody MenuItemsLayoutRequest request) {
        String applicationId = pageMenuService.getMenuApplicationId(id);
        SecurityUtils.requireServicePermission(ServicePermission.MENUS_MANAGE);
        SecurityUtils.requireApplicationAccess(applicationId);
        return ResponseEntity.ok(pageMenuService.updateMenuItemsLayout(id, request == null ? List.of() : request.getItems()));
    }

    @PutMapping("/{id}/items/{itemId}")
    public ResponseEntity<MenuResponse> updateItem(@PathVariable String id,
                                                   @PathVariable String itemId,
                                                   @Valid @RequestBody MenuItemUpsertRequest request) {
        String applicationId = pageMenuService.getMenuApplicationId(id);
        SecurityUtils.requireServicePermission(ServicePermission.MENUS_MANAGE);
        SecurityUtils.requireApplicationAccess(applicationId);
        return ResponseEntity.ok(pageMenuService.updateMenuItem(id, itemId, request));
    }

    @DeleteMapping("/{id}/items/{itemId}")
    public ResponseEntity<MenuResponse> deleteItem(@PathVariable String id, @PathVariable String itemId) {
        String applicationId = pageMenuService.getMenuApplicationId(id);
        SecurityUtils.requireServicePermission(ServicePermission.MENUS_MANAGE);
        SecurityUtils.requireApplicationAccess(applicationId);
        return ResponseEntity.ok(pageMenuService.deleteMenuItem(id, itemId));
    }
}

File: backend-nestjs/src/controllers/admin-menu.controller.ts
import { Body, Controller, Delete, Get, Param, Patch, Post, Put, Query, Req } from '@nestjs/common';
import type { Request } from 'express';
import { AdminAuthorizationService } from '../auth/admin-authorization.service';
import { ServicePermission } from '../auth/admin-permissions';
import { MenuStatus } from '../common/menu-types';
import { MenuItemUpsertRequestDto } from '../dto/requests/menu-item-upsert-request.dto';
import { MenuItemsLayoutRequestDto } from '../dto/requests/menu-items-layout-request.dto';
import { MenuStatusRequestDto } from '../dto/requests/menu-status-request.dto';
import { MenuUpsertRequestDto } from '../dto/requests/menu-upsert-request.dto';
import { MenuResponseDto } from '../dto/responses/menu-response.dto';
import { AdminMenuService, MenuContentCandidateDto } from '../services/admin-menu.service';
import { TenantRouteSyncRequestDto } from '../dto/requests/tenant-route-sync-request.dto';
import { TenantRouteService } from '../services/tenant-route.service';

@Controller('/api/v1/admin/menus')
export class AdminMenuController {
  constructor(
    private readonly menuService: AdminMenuService,
    private readonly access: AdminAuthorizationService,
    private readonly tenantRouteService: TenantRouteService,
  ) {}

  @Put('routes/sync')
  async syncTenantRoutes(
    @Req() request: Request,
    @Body() body: TenantRouteSyncRequestDto,
  ) {
    const applicationId = this.access.getApplicationId(request);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    return this.tenantRouteService.syncByApplicationId(applicationId, body);
  }

  @Post()
  async create(@Req() request: Request, @Body() body: MenuUpsertRequestDto): Promise<MenuResponseDto> {
    const applicationId = this.access.getApplicationId(request);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    return this.menuService.create({ ...body, applicationId });
  }

  @Put(':id')
  async update(@Req() request: Request, @Param('id') id: string, @Body() body: MenuUpsertRequestDto): Promise<MenuResponseDto> {
    const applicationId = await this.menuService.getApplicationIdById(id);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    return this.menuService.update(id, body);
  }

  @Patch(':id/status')
  async changeStatus(@Req() request: Request, @Param('id') id: string, @Body() body: MenuStatusRequestDto): Promise<MenuResponseDto> {
    const applicationId = await this.menuService.getApplicationIdById(id);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    return this.menuService.changeStatus(id, body.status);
  }

  @Delete(':id')
  async delete(@Req() request: Request, @Param('id') id: string): Promise<{ ok: boolean }> {
    const applicationId = await this.menuService.getApplicationIdById(id);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    await this.menuService.delete(id);
    return { ok: true };
  }

  @Get(':id')
  async get(@Req() request: Request, @Param('id') id: string): Promise<MenuResponseDto> {
    const applicationId = await this.menuService.getApplicationIdById(id);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    return this.menuService.get(id);
  }

  @Get(':id/published-content')
  async publishedContent(@Req() request: Request, @Param('id') id: string): Promise<MenuContentCandidateDto[]> {
    const applicationId = await this.menuService.getApplicationIdById(id);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    return this.menuService.listPublishedContentCandidates(id);
  }

  @Post(':id/sync-published')
  async syncPublished(@Req() request: Request, @Param('id') id: string): Promise<MenuResponseDto> {
    const applicationId = await this.menuService.getApplicationIdById(id);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    return this.menuService.syncPublishedContent(id);
  }

  @Get()
  async list(
    @Req() request: Request,
    @Query('languageCode') languageCode?: string,
    @Query('status') status?: MenuStatus,
  ): Promise<MenuResponseDto[]> {
    const applicationId = this.access.getApplicationId(request);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    return this.menuService.list(applicationId, languageCode, status);
  }

  @Post(':id/items')
  async addItem(@Req() request: Request, @Param('id') id: string, @Body() body: MenuItemUpsertRequestDto): Promise<MenuResponseDto> {
    const applicationId = await this.menuService.getApplicationIdById(id);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    return this.menuService.addItem(id, body);
  }

  @Put(':id/items/layout')
  async updateItemsLayout(
    @Req() request: Request,
    @Param('id') id: string,
    @Body() body: MenuItemsLayoutRequestDto,
  ): Promise<MenuResponseDto> {
    const applicationId = await this.menuService.getApplicationIdById(id);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    return this.menuService.updateItemsLayout(id, body.items ?? []);
  }

  @Put(':id/items/:itemId')
  async updateItem(
    @Req() request: Request,
    @Param('id') id: string,
    @Param('itemId') itemId: string,
    @Body() body: MenuItemUpsertRequestDto,
  ): Promise<MenuResponseDto> {
    const applicationId = await this.menuService.getApplicationIdById(id);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    return this.menuService.updateItem(id, itemId, body);
  }

  @Delete(':id/items/:itemId')
  async deleteItem(@Req() request: Request, @Param('id') id: string, @Param('itemId') itemId: string): Promise<MenuResponseDto> {
    const applicationId = await this.menuService.getApplicationIdById(id);
    this.access.assertServiceAccess(request, ServicePermission.MENUS_MANAGE, applicationId);
    return this.menuService.deleteItem(id, itemId);
  }
}
